### PR TITLE
Add handle validation pass to `Validator`

### LINKED ISSUE #1692: Handle validation pass
We should consider introducing an early pass to validation that checks all handles are valid, allowing the rest of validation to assume that arena indexing is infallible.

We recently improved the validator's robustness when presented with modules with invalid `Handle`s. This was necessary, but it had an unpleasant impact on the validator code and some interfaces. Some operations became fallible (e.g. [`try_size`](https://github.com/gfx-rs/naga/blob/f36cfb93dcd54ba1afea2b990b14e84982aeed5c/src/proc/mod.rs#L89), and various error types acquired bad-handle variants (e.g. [`TypeLayoutError::BadHandle`](https://github.com/gfx-rs/naga/blob/f36cfb93dcd54ba1afea2b990b14e84982aeed5c/src/proc/layouter.rs#L41)), and bits and pieces of [error-mapping code](https://github.com/gfx-rs/naga/blob/f36cfb93dcd54ba1afea2b990b14e84982aeed5c/src/proc/layouter.rs#L98) needed to be introduced.

Specifically:
- It's not always obvious where the checks are needed. We found most of the current set by fuzzing.
- Back ends are permitted to assume validation has succeeded, so they either need to unwrap (setting an uncomfortable precedent) or call alternate functions that panic on bad handles (boilerplate, and unclear when to use which alternative)
- The checks are a distraction from the code's primary purpose.
- There's a lot of redundancy in the error types.

An alternative would be to have one pass that runs at the beginning of validation that walks all functions, statement trees, expressions, types, and constants, and checks all handles. Then subsequent validation code can assume handles are valid. It can use the `Index` implementations on arenas, simplify function return types and error enums, and omit error remapping code.

The handle validation pass need not return very detailed error information; handle validation errors are only of interest to front end developers, not WebGPU or wgpu users. Capturing a backtrace would be the most important thing.

### PR DESCRIPTION
**Pull request by ErichDonGubler**
Resolves #1692. Recommended review experience is commit-by-commit.

### CODE REVIEW

**Comment by ErichDonGubler** [comment]
@jimblandy: I think that this is finally ready! 🙌🏻

**Comment by ErichDonGubler** [comment]
@jimblandy: Just wanted to ping here again, and get confirmation on whether reviewing this is a priority for you or not. Whatever the case, thanks in advance! 🙂

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:55]: Everything that applies to `Array` sizes applies here as well, I think.

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:99]: We need to check the `index` of an `Access` expression as well.

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:94]: Could we just inline this into `validate_function`?

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:162]: The way this turned out, I think it'd just be clearer to have the `match` on the outside, and call `check_dep_opt` from the `ImageQuery::Size` arm. Listing out all the variant is good, though.

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:233]: The arguments and result have type handles.

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:236]: These have expression handles.

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:237]: Since many kinds of statements have expression handles (`Statement::Emit`, `Statement::If`) and function handles (`Statement::Call`), we'll need a recursive function to walk the statement tree and check those handles.

File: src/valid/mod.rs
Comment (jimblandy) [inline src/valid/mod.rs:316]: I think this should go after the `reset` and `reset_types` calls, and be conditional on the `"validate"` feature. When that's turned off, panicking is just fine. :D

**Review by jimblandy — CHANGES_REQUESTED** [review]
Okay, this looks great - just what we need. There are a few omissions that need to be addressed, and I had some suggestions.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: I'm actually not sure about the balancing act WRT where to stop destructuring. I'm a fan of _destructuring all the things_, but it _does_ make quite a few humans recoil at the thought of “I have to change this _potentially every time_ I change this representation?” To me, though, it never _really_ causes significant problems; mostly unexpected compile failures that are easy, if not trivial, to resolve. LSP refactors + `rustc` make for a nice toolkit to quickly resolve compile errors, if even needed.

On the flip side, fuzzers are usually a (vital) last line of defense against open-ended input issues, but I also don't think the maintenance burden is a significant motivator for abdicating analytical defense-in-depth for this pass. However, I'm not a long-time maintainer of _any_ code like this, so I'll defer to your opinion here.

Comment (ErichDonGubler) [reply src/valid/handles.rs]: Destructured s'more with d48c7133, BTW. 🙂

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs:80]: RE: needing to do some low-key type checking here: it _does_ feel a little yucky to take on some overlap with type-checking, but I feel like that's the fundamental trade-off with having multiple passes, in general. I have implemented a fix with 468b72d5, taking the route of converting a crate-private error type to existing `ValidationError` variants (which is similar to your suggestion to implement `InvalidHandleError::InvalidArraySizeConstant`, but not quite the same); LMK what you think.

> The array size is a `Constant`, so that handle needs to be validated here.

Ah, so I missed the `BindingArray` ca—ah, and I see you've noted it already. 😅

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: Oh, oof, good eye. Fixed with 4c5e2099.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: Updated as suggested with 555f4207.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: Oh, oof, I feel silly for missing this. 😳 Resolved with 3fba607f05b3f958d45e6139453cbdb5240fd5e0.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: Oof again. 😖 Resolved with 9ce5abba.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: Yep, this was a pretty big omission on my part, sorry. 😓 Pushed 23637853 and 75654ac7. Whaddaya think?

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs:87]: I changed to destructure more; see also [this comment](https://github.com/gfx-rs/naga/pull/2090#discussion_r1047294339).

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs:35]: Renamed as suggested with fe7e59cb.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: Since you're using [a reference to another convo](https://github.com/gfx-rs/naga/pull/2090#discussion_r1044946235)...here's [a reference to my answer there](https://github.com/gfx-rs/naga/pull/2090#discussion_r1047294521). 😉

File: src/valid/mod.rs
Comment (ErichDonGubler) [reply src/valid/mod.rs]: Changed as suggested in 4b1165fb. I feel silly for not considering the question of `cfg`'ing things, given the surrounding code. 😓

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: ~~Sure! Changed with 3412651ac5ccfbf5a1a5e5158cf0c00760fe1ff9.~~ EDIT: Reverted with 202d699a06c6607c57980d727a0bc299b8a77caf.

Actually, the inner `match` here is still something we'll want to re-use. We can inline the looping logic (which I did with 9ce5abba). After moving the closure to an associated function (see 23637853), I did end up re-using it with 75654ac7 (see also the convo we have about statements not getting checked [here](https://github.com/gfx-rs/naga/pull/2090#discussion_r1047296957)).

File: src/proc/mod.rs
Comment (ErichDonGubler) [reply src/proc/mod.rs:104]: Huh. I would have thought that it would be _less_ clear what error handling to expect from what is now the only public way to get the size of a type. 🤔

In general, I've found it more beneficial to document the conditions under which a public API could panic. You're saying that you don't think that that outweighs the maintenance cost, in this case?

Comment (ErichDonGubler) [reply src/proc/mod.rs:104]: (OT, but still most relevant to this convo out of this whole PR: remove the `the` article with c533142a.)

Comment (ErichDonGubler) [reply src/proc/mod.rs]: I've removed the `Panics` section with 65251d65.

**Comment by ErichDonGubler** [comment]
Did some further `clippy` warning hunting with a02b02b8ca0a471c7ccc2c964a521920b5794b0f, 00f4f6b4, ce6f85e8, and 64932357.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: Fixed as suggested with e7cc0df4.

**Comment by ErichDonGubler** [comment]
@jimblandy: Once you've had a chance to look through the round of fixes I've pushed up here, LMK, and I'll squash things down into a sane set of commits again. Everything squashes down cleanly, so 😤.

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:536]: These comments about type-erased handles don't apply here.

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:543]: Suggest:

```suggestion
    // This error is used for many `Handle` types, but there's no point in making this generic,
    // so we just flatten them all to `Handle<()>` here.
    subject: Handle<()>,
    depends_on: Handle<()>,
```

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:551]: similar suggestion

**Review by jimblandy — CHANGES_REQUESTED** [review]
When our CI runs:
```
cargo check -p naga --no-default-features
```
it gets a lot of warnings about unused functions, etc. in `src/valid/handles.rs`. We should put `#[cfg(feature = "validate")]` in the appropriate places (perhaps on the entire `valid::handles` module?) to make the non-validating configuration compile cleanly again.

File: src/proc/mod.rs
Comment (jimblandy) [reply src/proc/mod.rs]: > In general, I've found it more beneficial to document the conditions under which a public API could panic. You're saying that you don't think that that outweighs the maintenance cost, in this case?

That's a fair question, and it's a point that comes up in many contexts, so let me explain my thinking:

The information content of a message is the -log₂ of the probability of its being received: for example, a message that is received with probability 0.25 conveys two bits of information. But this means that a message that is almost always received has a probability approaching 1, and thus conveys almost zero bits. Hearing "good morning" from someone who always says that is nice, but it doesn't tell you much.

Almost every public Naga API that operates on a module or a part of a module is going to access arenas, and thus will panic in the presence of invalid handles. If we were to place these panic notices correctly, almost every public API in Naga would have a "Panics" section saying the same thing. That would be a message with a probability approaching one, and thus one that conveys little information, and thus one that might be best omitted, since the reader has a limited attention span.

However, if we document this general approach in higher-level documentation, like the lib.rs comments on `Arena`s, that's being read by users who are getting an overview of Naga and are new to its API. Such readers are probably not ready to make many assumptions about Naga's behavior, so explaining that validation checks for bad handles, but subsequent stages assume these checks have been done and will panic (but not exhibit UB) if presented with bad handles, is not at all expected, and would convey many bits. So putting the message in a place where it has the most effect on the readers' understanding improves its information content.

The Rust standard library documentation for slice methods has a "Panics" section on every method that takes an index. For example, `split_at(&self, mid: usize)` says, "Panics if mid > len." I *think* this situation is different from Naga's, because there are various ways to respond to an index that is out of bounds. You could truncate (like the `windows` method when `size` is too large), it could be UB (like `split_unchecked`), etc. But in Naga, a bad handle is always meaningless. So there's really no other possible response.

In any case, if we're going to mention the panics, then I want to do it systematically - and that's not for this PR.

**Review by jimblandy — COMMENTED** [review]
I'll need to get back to the rest of this tomorrow, but here are some comments so far.

**Comment by ErichDonGubler** [comment]
> When our CI runs:
> 
> ```
> cargo check -p naga --no-default-features
> ```
> 
> it gets a lot of warnings about unused functions, etc. in `src/valid/handles.rs`. We should put `#[cfg(feature = "validate")]` in the appropriate places (perhaps on the entire `valid::handles` module?) to make the non-validating configuration compile cleanly again.

Silenced warnings with 2fab863a.

**Comment by ErichDonGubler** [comment]
@jimblandy:

> When our CI runs … it gets a lot of warnings about unused functions, etc. in `src/valid/handles.rs`. We should put `#[cfg(feature = "validate")]` in the appropriate places (perhaps on the entire `valid::handles` module?) to make the non-validating configuration compile cleanly again.

Silenced warnings with 113957aa.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs:524]: How's 7f0faf9d look?

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: N.B. that the `From` conversions here may not be necessary with some well-placed `map_err`; I hadn't tried that, but that could cut the trait implementations down, in case that changes your calculus.

WRT moving to `ValidationError`, I'm not feeling picky, I guess. Moved to "just" return a `ValidationError` with df755c36.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: Fixed with 53eca848.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: Fixed with 53eca848.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: Fixed with 53eca848.

**Comment by ErichDonGubler** [comment]
I've autosquashed all reviewed content into a few (hopefully nice) commits. Now that I've rebased onto `master`, I'm getting unrelated [CI failures](https://github.com/gfx-rs/naga/actions/runs/3707003285/jobs/6282899141). I've submitted https://github.com/gfx-rs/naga/pull/2176 to resolve them.

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:97]: Suggestion: combine this `match` arm with with `TypeInner::Array` via an `|` pattern.

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:69]: It's Naga style to match `*inner`, drop the references from all the patterns, and use `ref` as appropriate to avoid moves. This is Clippy lint [match_ref_pats](https://rust-lang.github.io/rust-clippy/master/index.html#match_ref_pats), so I don't know why Clippy isn't complaining about this.

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:385]: Similarly: please `match *stmt` and simplify all the patterns.

File: src/valid/handles.rs
Comment (jimblandy) [inline src/valid/handles.rs:273]: The `level` field is a `SampleLevel`, which can have handles.

**Review by jimblandy — CHANGES_REQUESTED** [review]

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: If `clippy` doesn't or can't enforce this style…is it worth enforcing manually? 🤔

Changed as suggested with af211507.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: Changed as suggested with 62f43a66.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs:270]: Gonna need to make some pizza…with all this d'oh. 😩

Resolved with e1229f1b.

File: src/valid/handles.rs
Comment (ErichDonGubler) [reply src/valid/handles.rs]: Changed as suggested with b43468a6.

File: src/arena.rs
Comment (jimblandy) [inline src/arena.rs:130]: Since we've already added a `check_contains_handle` method to `Arena`, let's also add a `check_contains_range` method, let `valid::handles`' `Range::check_valid_for` method can just call that, and let this stay private.

File: src/valid/expression.rs
Comment (jimblandy) [inline src/valid/expression.rs:142]:
```suggestion
            // `Validator::validate_module_handles` should have caught this.
            panic!(
```

**Review by jimblandy — CHANGES_REQUESTED** [review]
Looks great! Just two small changes, and then I think we're ready to go.

File: src/valid/handles.rs
Comment (jimblandy) [reply src/valid/handles.rs]: It's all good, that's what reviews are for. (Pizza is tasty.)

**Review by jimblandy — APPROVED** [review]

**Pull request merged**
